### ISSUE TITLE
Perf Discussion -- [original title: Support `PRAGMA journal_mode=WAL;`?]

### ISSUE THREAD

**Issue by AlexErrant** [vlcn-io/js#27]
Using the latest https://github.com/vlcn-io/live-examples as my playground, I added

```ts
  console.log(await db.execO("PRAGMA journal_mode=memory;"));
```

which correctly yields `[
    {
        "journal_mode": "memory"
    }
]`

However,

```ts
console.log(await db.execO("PRAGMA journal_mode=WAL;"));
```

yields `[
    {
        "journal_mode": "delete"
    }
]`

Notably, `PRAGMA journal_mode=WAL;` works on https://rhashimoto.github.io/wa-sqlite/demo/

If it matters, I'm just trying WAL mode because it apparently may improve perf.

**Comment by AlexErrant** [comment]
Hm, I don't really follow the first part of your comment; I'm not currently using reactive/live queries (though perhaps in the future! I'm using SolidJS, after all.) 

In my app there's [an MS Excel-style spreadsheet](https://github.com/AlexErrant/Pentive/blob/main/app/src/components/cardsTable.tsx) that may have 10k+ rows that a user may search/sort on; so we virtualize it and query for data [in blocks of 100 rows](https://github.com/AlexErrant/Pentive/blob/96cc294f3a391c0c37f600a13549b9f77a69dfa6/app/src/components/cardsTable.tsx#L242). Being Excel-style, you can arbitrarily scroll any distance down the spreadsheet. `OFFSET` has terrible perf when skipping 15k+ rows, and users might want to jump to the bottom of the spreadsheet. I handle this in two steps:

1. Return the first 100 results of a query ASAP.
2. [In a nonblocking manner](https://github.com/AlexErrant/Pentive/blob/96cc294f3a391c0c37f600a13549b9f77a69dfa6/app/src/sqlite/card.ts#L387), add [_all_ the relevant ids for that query to a temp table](https://github.com/AlexErrant/Pentive/blob/96cc294f3a391c0c37f600a13549b9f77a69dfa6/app/src/sqlite/card.ts#L135).

When a user tires of the first 100 results and starts randomly scrolling, I can use cursor pagination with the temp table's rowid (`WHERE rowid > 15000 ORDER BY rowid LIMIT 100`) by taking advantage of the fact that the spreadsheet's row number is now the same as the temp table's rowid. Building this cache is relatively expensive; it takes ~10 seconds on my devbox with ~15k rows. However jumping to the bottom of that result set takes ~200ms, so I'm happy with it.

The problem occurs when a user sees something interesting in the first 100 results and clicks it, generating a read query to get details. _The cache write is blocking the detail query._ I wondered if WAL mode would be an easy way of getting concurrent read/writes, hence this issue.

This isn't blocking me, so please don't feel obliged to make this issue any kind of priority; there's a reason I closed it :) A workaround is to build the cache _incrementally_... which I'll happily procrastinate on until I hate myself enough to take action (or W3C TAG solves my problem, which I give a 0.000001% chance of happening).

**Comment by rhashimoto** [comment]
> The problem occurs when a user sees something interesting in the first 100 results and clicks it, generating a read query to get details. The cache write is blocking the detail query. I wondered if WAL mode would be an easy way of getting concurrent read/writes, hence this issue.

I don't think that WAL mode would be a big win here. IIUC both your cache query and your user-initiated follow-up query only read from the main database, so this usage only benefits from concurrent reads, not concurrent reads and writes. Writing to a temp table doesn't change this.

You can get concurrent reads with wa-sqlite, but two things need to be true, both of which I'm guessing are currently not true:

1. You must execute each read transaction from a separate WebAssembly instance, i.e. wa-sqlite must be initialized multiple times, ideally using separate JavaScript contexts. This is because WASM SQLite doesn't allow multiple calls to be in flight simultaneously in the same WASM instance.
2. You must be using an IndexedDB VFS that supports shared read locks. I believe that vlcn currently uses the stock IDBBatchAtomicVFS class, which uses only exclusive locks.

Your plan to build a cache with multiple smaller queries that you can preempt sounds like a reasonable alternative approach.

Of course, the catch-all tactic to improving SQLite performance is to [increase the page cache size](https://www.sqlite.org/pragma.html#pragma_cache_size) if you haven't done that already.

@tantaman Does cr-sqlite provide a way to stream query result rows? That seems like a generally useful feature in a higher level API.

**Comment by rhashimoto** [comment]
> I currently don't expose step but that would be a good thing to expose.

Exposing step is sort of a lower level approach. I was thinking that supplying a ReadableStream source, i.e. the object you pass to a [ReadableStream constructor](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/ReadableStream), for output rows might be a nice way to go.

The only API you would need to define is how to create the source - after that everything is specified by the standard Streams API. So there's not much external design and documentation to do, and then you get Streams API features for free, like buffering, back pressure, and async iteration.
